src/sql/src/plan/statement/ddl.rs

```diff
@@ -2995,97 +2995,205 @@ fn kafka_sink_builder(
+        }
 
-            if key_desc_and_indices.is_some()
-                && (avro_key_fullname.is_some() ^ avro_value_fullname.is_some())
-            {
-                sql_bail!("Must specify both AVRO KEY FULLNAME and AVRO VALUE FULLNAME when specifying generated schema names");
-            }
+        let options = AvroSchemaOptions {
+            avro_key_fullname,
+            avro_value_fullname,
+            set_null_defaults: null_defaults,
+            is_debezium: matches!(envelope, SinkEnvelope::Debezium),
+            sink_from: Some(sink_from),
+            value_doc_options,
+            key_doc_options,
+        };
+        Ok((
+            csr_connection,
+            options,
+            key_compatibility_level,
+            value_compatibility_level,
+        ))
+    };
 
-            let options = AvroSchemaOptions {
-                avro_key_fullname,
-                avro_value_fullname,
-                set_null_defaults: null_defaults,
-                is_debezium: matches!(envelope, SinkEnvelope::Debezium),
-                sink_from: Some(sink_from),
-                value_doc_options,
-                key_doc_options,
-            };
+    // Helper function to determine if a relation desc is valid for text or binary format.
+    // We only allow a single column of non-list/map type.
+    let valid_text_binary_desc = |desc: &RelationDesc| {
+        if desc.arity() != 1 {
+            false
+        } else if desc.iter().any(|(_, ty)| {
+            matches!(
+                ty.scalar_type,
+                ScalarType::List { .. } | ScalarType::Map { .. }
+            )
+        }) {
+            false
+        } else {
+            true
+        }
+    };
 
+    let map_format_non_avro = |format: &Format<Aug>, desc: &RelationDesc| match format {
+        Format::Json { array: false } => Ok::<_, PlanError>(KafkaSinkFormatType::Json),
+        Format::Bytes if valid_text_binary_desc(desc) => Ok(KafkaSinkFormatType::Bytes),
+        Format::Text if valid_text_binary_desc(desc) => Ok(KafkaSinkFormatType::Text),
+        Format::Bytes | Format::Text => {
+            bail_unsupported!("BYTE/TEXT format with multiple columns or non-scalar types")
+        }
+        Format::Json { array: true } => bail_unsupported!("JSON ARRAY format in sinks"),
+        format => bail_unsupported!(format!("sink format {:?}", format)),
+    };
+
+    // Map from the format specifier of the statement to the individual key/value formats for the sink.
+    let format = match format {
+        Some(FormatSpecifier::KeyValue {
+            key: Format::Avro(..),
+            value: Format::Avro(..),
+        }) => {
+            // Since Avro format requires specifying CSR connection options, if we allow both key and value
+            // to be separately specified as Avro we would need to deconflict the CSR connection options
+            // between them, so we disallow this.
```

Comment:
Is there a problem with specifying two separate CSR connections? We can publish to whatever the user instructs us to (see other comment)

Reply:
There isn't - I'll refactor to allow it